src/include/txn.i

```diff
@@ -1088,18 +1103,28 @@ __wt_txn_update_check(WT_SESSION_IMPL *session, WT_UPDATE *upd)
     ignore_prepare_set = F_ISSET(txn, WT_TXN_IGNORE_PREPARE);
     F_CLR(txn, WT_TXN_IGNORE_PREPARE);
     for (; upd != NULL && !__wt_txn_upd_visible(session, upd); upd = upd->next) {
-        if (upd->txnid != WT_TXN_ABORTED) {
-            if (ignore_prepare_set)
-                F_SET(txn, WT_TXN_IGNORE_PREPARE);
-            WT_STAT_CONN_INCR(session, txn_update_conflict);
-            WT_STAT_DATA_INCR(session, txn_update_conflict);
-            return (__wt_txn_rollback_required(session, "conflict between concurrent operations"));
-        }
+        if (upd->txnid != WT_TXN_ABORTED)
+            WT_ERR(__txn_rollback_required(session));
+    }
+
+    /* If there is no cursor, we don't need to check the on page value. */
+    WT_ASSERT(session, cbt != NULL || !check_onpage_value);
+
+    /* Check conflict against the on page value. */
+    if (upd == NULL && check_onpage_value) {
```

Comment:
Another thing I missed. We only need to check the on page value if nothing is on the update chain except aborted updates.